crates/sparrow-scheduler/src/task.rs

```diff
@@ -0,0 +1,140 @@
+use std::sync::Arc;
+
+use error_stack::ResultExt;
+use sparrow_arrow::Batch;
+
+use crate::queue::Queue;
+use crate::schedule_count::ScheduleCount;
+use crate::{Error, Partition, Pipeline, Worker};
+
+/// The unit of work executed by the scheduler.
+///
+/// A task processes a single unit of input (typically a batch), applies a
+/// single [Pipeline] and produces a single unit of output (typically a batch).
+#[derive(Debug)]
+pub struct Task<T: Pipeline> {
+    /// Index of this pipeline. Used for debugging.
+    index: usize,
+    /// The pipeline to execute.
+    ///
+    /// This is a weak reference to avoid cycles.
+    pipeline: std::sync::Weak<T>,
+    /// The partition of the pipeline to execute.
+    partition: Partition,
+    /// An atomic counter tracking how many times the task has been submitted.
+    ///
+    /// This is reset after the task is executed.
+    schedule_count: ScheduleCount,
+}
+
+impl<T: Pipeline> Task<T> {
+    /// Create a new task executing the given pipeline and partition.
+    pub(crate) fn new(index: usize, pipeline: std::sync::Weak<T>, partition: Partition) -> Self {
+        Self {
+            index,
+            pipeline,
+            partition,
+            schedule_count: ScheduleCount::default(),
+        }
+    }
+
+    fn pipeline(&self) -> error_stack::Result<Arc<T>, Error> {
+        Ok(self
+            .pipeline
+            .upgrade()
+            .ok_or_else(|| Error::PipelineDropped {
+                index: self.index,
+                name: T::name(),
+                partition: self.partition,
+            })?)
+    }
+
+    fn error(&self, method: &'static str) -> Error {
+        Error::Pipeline {
+            method,
+            index: self.index,
+            name: T::name(),
+            partition: self.partition,
+        }
+    }
+}
+
+/// Trait with methods on generic `Task`.
+///
+/// This is an object safe wrapper allowing execution of the pipeline within the task.
+pub trait TaskObject: std::fmt::Debug + Sync + Send {
+    /// Record a request for scheduling.
```

Comment:
Does it return `false` if it was already scheduled -- are there situations when it is not schedulable?

Reply:
This is related to re-entrancy.

If we schedule "run partition 1 of this transform pipeline on batch 1" and it isn't currently scheduled, then we need to add it to a local queue. If we then get *another* request to "run partition 1 of this transform pipeline on batch 2", we don't want to schedule it again (since that could lead to running it on two threads). The first time this returns "true" (indicating we should schedule that pipeline/partition) the second time it returns "false" (indicating we already have it in a queue).

Then, while we're running "partition 1 of this transform pipeline on batch 1", if we call this it will return false (causing us not to immediately schedule it), but while we're finishing we can detect "this partition was scheduled while we were running it, so add it to the queue again".

Reply:
Updated wording.